### DISCUSSION TITLE
Redirects don't change to GET method

### DISCUSSION THREAD

**Discussion by johanrosenson** [Kong/insomnia#3753] [Ideas]
**Describe the bug**
When following redirects is active the redirect will not change to using the GET method as it should, but will keep the original request method.

**To Reproduce**
1. Make a POST request to a URL that responds with a redirect (302).
2. Switch to timeline and see that the redirect did not switch to GET.

I managed to reproduce with a POST to http://httpbin.org/redirect-to?url=http%3A%2F%2Fexample.com%2F

(Same example as in previous issues linked below)

**Expected behavior**
The redirect should switch to using GET.

**Screenshots**
<img width="465" alt="Screenshot 2020-05-16 at 11 57 10" src="https://user-images.githubusercontent.com/10432296/82111067-6aa9da80-976c-11ea-8388-6cb9cc92ef97.png">

**Desktop (please complete the following information):**
 - OS: macOS 10.15.4
 - Installation Method: homebrew
 - App Version 7.1.1

**Additional context**
I have only tested this using a POST request, cannot confirm or rule out that it also happens with PUT, PATCH and other methods.

Previous issues with the same bug:
https://github.com/Kong/insomnia/issues/227
https://github.com/Kong/insomnia/issues/439

**Comment by nijikokun** [comment]
@johanrosenson Thanks for reporting.

I'm just learning about this subject so... reading the RFC there is a User Agent required to specify that the method should be changed to GET.

In your Response Headers on the 301, I'm not seeing a user-agent specifying as such. That said, it doesn't look like CURL even supports this 🤦

[MDN:](https://developer.mozilla.org/en-US/docs/Web/HTTP/Redirections#attr1)

> The specification did not intend to allow method changes, but there are existing user agents that do change their method.

[RFC7231:](https://tools.ietf.org/html/rfc7231#section-6.4.2) which doesn't denote what it looks like 🤦‍♀️

> Note: For historical reasons, a user agent MAY change the request method from POST to GET for the subsequent request. 

and from [CURLOPT_POSTREDIR](https://curl.haxx.se/libcurl/c/CURLOPT_POSTREDIR.html):

> The non-RFC behaviour is ubiquitous in web browsers, so the library does the conversion by default to maintain consistency.

Seems like maybe we need some tests surrounding this. I've labeled it as accepted, since it appears to be recurring and some research seems to be required.

It might have to do something with the introduction of [CUSTOMREQUEST](https://curl.haxx.se/libcurl/c/CURLOPT_CUSTOMREQUEST.html) in #377.

**Comment by johanrosenson** [comment]
From your link:

> The specification did not intend to allow method changes, but there are existing user agents that do change their method. 308 was created to remove the ambiguity of the behavior when using non-GET methods.

My experience is that the real world usage of 301 in all browsers is to always change to a GET, I think Insomnia need to follow suit and do the same (no matter what the specification originally intended), otherwise Insomnia would differ from all the browsers, which does made some testing more difficult then it needs to be.

It also says that 308 was created for this very reason; to remove the confusion.
So the real world usage is clearly to change to a GET for 301 (including 302 and 303), 307 and 308 are clearer in their specification and do not change the method when redirecting.

This problem have already have been an issue twice before (and twice fixed?), but i think a third fix may be needed.

Thanks for accepting the issue.

**Comment by mlh758** [comment]
I'm using Insomnia 2020.3.3 and I've observed this behavior:

If using multipart/form-data as the content type with a POST request it will follow the redirect and send POST again. If using application/json Insomnia will follow the redirect and change to a GET request. My version of curl:

```
curl 7.64.1 (x86_64-apple-darwin19.0) libcurl/7.64.1 (SecureTransport) LibreSSL/2.8.3 zlib/1.2.11 nghttp2/1.39.2
Release-Date: 2019-03-27
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtsp smb smbs smtp smtps telnet tftp
Features: AsynchDNS GSS-API HTTP2 HTTPS-proxy IPv6 Kerberos Largefile libz MultiSSL NTLM NTLM_WB SPNEGO SSL UnixSockets
```

just doesn't follow the redirect at all with a POST request.

**Comment by MagicShoebox** [comment]
Also ran into this, see https://github.com/curl/curl/issues/5145 for more details.
TL:DR: the `-X/--request` flag **overrides** the switch to GET. In fact, if run with `-v`, the curl command line will log "Switching to GET" and then send off a second POST anyway. Removing the `-X/--request` flag and relying on `--data` to implicitly set it works as expected. I, for one, did not find this intuitive.

I don't know exactly how Insomnia interacts with curl/libcurl, but I think there's a good chance it explicitly sets the method, which curl then uses to override the switch to GET as detailed in the other thread.

Some speculation, but I hope it helps someone!

**Comment by draconteus** [comment]
I'm using this version:

```
Version: Insomnia Core 2020.4.1
Release date: 23/09/2020
OS: Darwin x64 19.6.0
Electron: 9.1.1
Node: 12.14.1
V8: 8.3.110.13-electron.0
Architecture: x64
node-libcurl: libcurl/7.69.1 OpenSSL/1.1.1g zlib/1.2.11 brotli/1.0.7 libidn2/2.1.1 libssh2/1.9.0 nghttp2/1.41.0
```


The POST request (with JSON payload) actually switched to GET (and this is a problem for us), but PUT/PATCH/etc. is not.

I would like to have the option to enforce keeping the original request method as it can be a valid use case (api is behind Cloudflare which responds with 301). Also it's an option for curl (using `-X/--request` explicitly) or even in Postman.

**Comment by wdawson** [comment]
Hey everyone. As several of you mentioned, there are needs on both sides of this (keeping the method the same and changing it). We want Insomnia to be flexible and allow you to make any kind of request that you need it to. I've moved this to a discussion to fit in with other product ideas. Thanks for bringing this up!

**Reply by zapp-b** [reply]
Is there a current solution that allows you to switch between the two?

**Comment by fxjordan** [comment]
Hey, I just wanted to mention that this behavior can **really** be a problem for some people.

I was executing a `DELETE` request (for a vote resource) which worked 100 times in the JavaScript frontend. The API always responds with a 303 to a parent resource, which I didn't know
Now, after I executed the request in Insomnia, the `DELETE` was executed for that parent resource too, which accidentally deleted the whole resource! Luckily, I was on a testing environment!

To avoid unexpected data-loss, I could think of the following changes:
- Open a dialog the first time a user receives a 303 to a `DELETE`, and asks the user if they want to proceed (according to the standard) and remind them to add a `User-Agent` if they want other behavior.
- Add the possibility to hide this warning dialog for the future
